### PR TITLE
ENT-210: Remove unnecessary pipeline element; allow ECU creation when consent is disabled

### PR DESCRIPTION
**Pull request by haikuginger**
**Description:** Previously, users would not be linked as EnterpriseCustomerUsers when data sharing consent was disabled. This PR remedies that, and also removes a third-party auth pipeline element that is no longer necessary due to previous changes.

**Sandbox URL:** https://pr14567.sandbox.opencraft.hosting (admin user: staff/edx)

**JIRA:** [ENT-210](https://openedx.atlassian.net/browse/ENT-210)

**Dependencies:** Requires edx/edx-platform#14567

**Merge deadline:** 13 March 2017

**Installation instructions:**

1. Set up a devstack with the branch in the linked edx-platform PR.
2. Set up a TestShib SAML integration with the instructions [found here](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/tpa/tpa_SAML_SP.html).
3. Create an EnterpriseCustomer in the Django admin.

**Testing instructions:**

1. Link the EnterpriseCustomer to the TestShib SAML integration
2. Verify that data sharing consent is set to disabled on the EnterpriseCustomer
3. In an incognito window, browse to the LMS registration page
4. Attempt to register via TestShib SAML
5. After registration, observe that you are logged in directly.
6. In the Django Admin panel for the EnterpriseCustomer, click Manage Learners
7. Observe that your SAML-created user is linked as an Enterprise Learner.

@pomegranited for initial review

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=53"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

File: enterprise/tpa_pipeline.py [hunk 1]
```diff
@@ -127,6 +127,10 @@ def redirect_to_consent():
         return
 
     if not customer.requests_data_sharing_consent:
```
Comment (mattdrayer) [inline enterprise/tpa_pipeline.py]: can we rename this to `enterprise_customer`?

File: enterprise/tpa_pipeline.py [hunk 2]
```diff
@@ -105,7 +105,7 @@ def get_consent_status_for_pipeline(pipeline):
 
 
 @partial
-def verify_data_sharing_consent(backend, user, **kwargs):
+def handle_enterprise_logistration(backend, user, **kwargs):
     """
     Verify that the data sharing consent state matches what's required.
```
Comment (mattdrayer) [inline enterprise/tpa_pipeline.py]: Let's update the docstring to reflect the correct format while we're here (Args, Returns)

File: enterprise/tpa_pipeline.py (hunk 1 above)
Comment (mattdrayer) [inline enterprise/tpa_pipeline.py]: Also please add a code comment describing the scenario covered by this block

File: tests/test_tpa_pipeline.py [hunk 3]
```diff
@@ -127,75 +127,37 @@ def test_get_consent_status(self):
             consent.delete()
             ec_user.delete()
 
-    def test_verify_data_sharing_consent(self):
+    def test_handle_enterprise_logistration(self):
         """
-        Test that we correctly verify consent status.
+        Test that we correctly handle incoming enterprise logistration requests.
```
Comment (mattdrayer) [inline tests/test_tpa_pipeline.py]: This docstring isn't really helpful -- it basically just restates the name of the test.

**Comment by mattdrayer** [comment]
Overall the change looks good -- I've basically just asked for some improved readability and additional context through better docstrings + code comments.

File: tests/test_tpa_pipeline.py [hunk 4]
```diff
@@ -127,75 +127,37 @@ def test_get_consent_status(self):
...
             )
             consent = UserDataSharingConsentAudit.objects.create(  # pylint: disable=no-member
                 user=ec_user
             )
-            assert isinstance(verify_data_sharing_consent(backend, self.user), HttpResponseRedirect)
+            assert isinstance(handle_enterprise_logistration(backend, self.user), HttpResponseRedirect)
             consent.state = 'enabled'
             consent.save()
-            assert verify_data_sharing_consent(backend, self.user) is None
+            assert handle_enterprise_logistration(backend, self.user) is None
```
Comment (pomegranited) [inline tests/test_tpa_pipeline.py]: There's two implicit cases not covered by these tests:
* `self.customer.enforce_data_sharing_consent == EnterpriseCustomer.DATA_CONSENT_OPTIONAL`
* `self.customer.enforce_data_sharing_consent == EnterpriseCustomer.AT_ENROLLMENT`

File: tests/test_tpa_pipeline.py [hunk 5]
```diff
@@ -127,75 +127,37 @@ def test_get_consent_status(self):
...
+            assert EnterpriseCustomerUser.objects.filter(
+                enterprise_customer=enterprise_customer,
+                user_id=self.user.id
+            ).count() == 1
         with mock.patch('enterprise.tpa_pipeline.get_ec_for_running_pipeline') as fake_get_ec:
             fake_get_ec.return_value = self.customer  # pylint: disable=redefined-variable-type
-            assert isinstance(verify_data_sharing_consent(backend, self.user), HttpResponseRedirect)
+            assert isinstance(handle_enterprise_logistration(backend, self.user), HttpResponseRedirect)
             ec_user = EnterpriseCustomerUser.objects.create(
```
Comment (pomegranited) [inline tests/test_tpa_pipeline.py]: Nit: This pattern of creating `ec_user` and `consent` objects, and then deleting them when the test is done, is used here and in the test above, [`test_get_consent_status`](https://github.com/open-craft/edx-enterprise/blob/ec78660e56416173846b7873ef852a7ba68b3463/tests/test_tpa_pipeline.py#L119).

Would be worth reducing some code duplication by pulling the creation out into a separate method, which also appends cleanup tasks to delete them when done ([example](https://github.com/edx/edx-platform/blob/6d2468d0dcf782a7561ecc61946146d6fae2762f/common/djangoapps/third_party_auth/tests/specs/test_twitter.py#L21-L28)), like this:

```python
    def create_ec_user_with_consent(self):
        ec_user = EnterpriseCustomerUser.objects.create(
            user_id=self.user.id,  # pylint: disable=no-member
            enterprise_customer=self.customer
        )
        consent = UserDataSharingConsentAudit.objects.create(  # pylint: disable=no-member
            user=ec_user
        )
        self.addCleanup(consent.delete)
        self.addCleanup(ec_user.delete)
        return ec_user, consent
```

File: tests/test_tpa_pipeline.py (hunk 4 above)
Comment (haikuginger) [inline tests/test_tpa_pipeline.py]: Good catch!

File: tests/test_tpa_pipeline.py (hunk 5 above)
Comment (haikuginger) [inline tests/test_tpa_pipeline.py]: So, it turns out the Django test suite does this for free. So I've removed the `.delete()` calls. As for creation, the Factory method is about as good as we can get. :)

**Comment by haikuginger** [comment]
@mattdrayer, @pomegranited, I've addressed your requests - a formal 👍 would be lovely. @saleem-latif, @asadiqbal08, a review from one of you is needed.

**Comment by mattdrayer** [comment]
And you shall have it!  Changes look great -- thanks very much for adding all the code comments -- 👍

File: tests/test_tpa_pipeline.py [hunk 6]
```diff
@@ -124,30 +124,69 @@ def test_get_consent_status(self):
...
+            fake_get_ec.return_value = self.customer
             assert isinstance(handle_enterprise_logistration(backend, self.user), HttpResponseRedirect)
+
+    def test_handle_enterprise_logistration_consent_optional(self):
+        """
+        Test that when consent is required, we redirect to the consent page.
+        """
+        backend = mock.MagicMock(name=None)
+        with mock.patch('enterprise.tpa_pipeline.get_ec_for_running_pipeline') as fake_get_ec:
+            self.customer.enforce_data_sharing_consent = EnterpriseCustomer.AT_ENROLLMENT
```
Comment (pomegranited) [inline tests/test_tpa_pipeline.py]: nit: This test is called `test_handle_enterprise_logistration_consent_optional`, but it tests consent `AT_ENROLLMENT`.

Switch the name with [L173](https://github.com/edx/edx-enterprise/pull/53/commits/4ad7791f8ce300c5f9c0201b59018acacfcefa29#diff-e668e167d9d0bf0073a948bc1c155a90R173) below.

File: tests/test_tpa_pipeline.py [hunk 7]
```diff
@@ -124,30 +124,69 @@ def test_get_consent_status(self):
...
+        """
+        backend = mock.MagicMock(name=None)
         with mock.patch('enterprise.tpa_pipeline.get_ec_for_running_pipeline') as fake_get_ec:
-            fake_get_ec.return_value = self.customer  # pylint: disable=redefined-variable-type
+            fake_get_ec.return_value = self.customer
             assert isinstance(handle_enterprise_logistration(backend, self.user), HttpResponseRedirect)
+
+    def test_handle_enterprise_logistration_consent_optional(self):
+        """
+        Test that when consent is required, we redirect to the consent page.
```
Comment (pomegranited) [inline tests/test_tpa_pipeline.py]: Same comment as 3 other tests; please distinguish this test.

File: tests/test_tpa_pipeline.py [hunk 8]
```diff
@@ -124,30 +124,69 @@ def test_get_consent_status(self):
...
+        """
+        backend = mock.MagicMock(name=None)
+        with mock.patch('enterprise.tpa_pipeline.get_ec_for_running_pipeline') as fake_get_ec:
+            self.customer.enforce_data_sharing_consent = EnterpriseCustomer.AT_ENROLLMENT
+            fake_get_ec.return_value = self.customer
+            assert isinstance(handle_enterprise_logistration(backend, self.user), HttpResponseRedirect)
+
+    def test_handle_enterprise_logistration_consent_required_at_enrollment(self):
+        """
+        Test that when consent is required, we redirect to the consent page.
```
Comment (pomegranited) [inline tests/test_tpa_pipeline.py]: Same comment as 3 other tests; please distinguish this test.

**Comment by haikuginger** [comment]
@pomegranited, changes made.

**Comment by haikuginger** [comment]
Rebased and added finishing touches from 53c1d66.

**Pull request merged**